File: ingestor/internal/mlb/schedule.go
package mlb

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const scheduleAPIBase = "https://statsapi.mlb.com/api/v1"

// ScheduleGame is metadata for one game returned by the MLB schedule endpoint.
type ScheduleGame struct {
	GamePK           int
	GameDate         string
	Season           int
	GameType         string
	Status           string
	StatusDetail     string
	AwayTeamID       int
	AwayTeamName     string
	AwayTeamAbbrev   string
	HomeTeamID       int
	HomeTeamName     string
	HomeTeamAbbrev   string
	AwayScore        *int
	HomeScore        *int
	VenueID          *int
	VenueName        string
	OfficialDate     string
}

// scheduleResponse is a minimal decode of GET /schedule for game discovery.
type scheduleResponse struct {
	Dates []struct {
		Games []scheduleGameRaw `json:"games"`
	} `json:"dates"`
}

type scheduleGameRaw struct {
	GamePK       int    `json:"gamePk"`
	GameDate     string `json:"gameDate"`
	OfficialDate string `json:"officialDate"`
	Season       string `json:"season"`
	GameType     string `json:"gameType"`
	Status       struct {
		AbstractGameState string `json:"abstractGameState"`
		DetailedState     string `json:"detailedState"`
	} `json:"status"`
	Teams struct {
		Away scheduleTeamSide `json:"away"`
		Home scheduleTeamSide `json:"home"`
	} `json:"teams"`
	Venue *struct {
		ID   int    `json:"id"`
		Name string `json:"name"`
	} `json:"venue"`
}

type scheduleTeamSide struct {
	Score *int `json:"score"`
	Team  struct {
		ID           int    `json:"id"`
		Name         string `json:"name"`
		Abbreviation string `json:"abbreviation"`
	} `json:"team"`
}

// ScheduleDateET returns today's date in YYYY-MM-DD for the America/New_York
// calendar, which MLB uses for schedule queries.
func ScheduleDateET(t time.Time) string {
	loc, err := time.LoadLocation("America/New_York")
	if err != nil {
		loc = time.UTC
	}
	return t.In(loc).Format("2006-01-02")
}

// FetchScheduleGames returns all regular-season games on the given calendar date.
func FetchScheduleGames(ctx context.Context, httpClient *http.Client, date string) ([]ScheduleGame, error) {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	if date == "" {
		date = ScheduleDateET(time.Now())
	}

	endpoint, err := url.Parse(scheduleAPIBase + "/schedule")
	if err != nil {
		return nil, fmt.Errorf("parse schedule url: %w", err)
	}
	q := endpoint.Query()
	q.Set("sportId", "1")
	q.Set("date", date)
	q.Set("gameTypes", "R")
	q.Set("hydrate", "team,linescore,venue")
	endpoint.RawQuery = q.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("build schedule request: %w", err)
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("fetch schedule: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return nil, fmt.Errorf("schedule status %d: %s", resp.StatusCode, strings.TrimSpace(string(body)))
	}

	var payload scheduleResponse
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return nil, fmt.Errorf("decode schedule: %w", err)
	}

	games := make([]ScheduleGame, 0)
	for _, day := range payload.Dates {
		for _, raw := range day.Games {
			games = append(games, mapScheduleGame(raw))
		}
	}

	return games, nil
}

func mapScheduleGame(raw scheduleGameRaw) ScheduleGame {
	gameDate := raw.OfficialDate
	if gameDate == "" && len(raw.GameDate) >= 10 {
		gameDate = raw.GameDate[:10]
	}

	season, _ := strconv.Atoi(raw.Season)

	gameType := raw.GameType
	if gameType == "" {
		gameType = "R"
	}

	game := ScheduleGame{
		GamePK:         raw.GamePK,
		GameDate:       gameDate,
		Season:         season,
		GameType:       gameType,
		Status:         raw.Status.AbstractGameState,
		StatusDetail:   raw.Status.DetailedState,
		AwayTeamID:     raw.Teams.Away.Team.ID,
		AwayTeamName:   raw.Teams.Away.Team.Name,
		AwayTeamAbbrev: raw.Teams.Away.Team.Abbreviation,
		HomeTeamID:     raw.Teams.Home.Team.ID,
		HomeTeamName:   raw.Teams.Home.Team.Name,
		HomeTeamAbbrev: raw.Teams.Home.Team.Abbreviation,
		AwayScore:      raw.Teams.Away.Score,
		HomeScore:      raw.Teams.Home.Score,
		OfficialDate:   gameDate,
	}

	if raw.Venue != nil {
		id := raw.Venue.ID
		game.VenueID = &id
		game.VenueName = raw.Venue.Name
	}

	return game
}

// IsLiveStatus reports whether MLB's abstract game state indicates active play.
func IsLiveStatus(status string) bool {
	switch strings.ToLower(strings.TrimSpace(status)) {
	case "live", "in progress":
		return true
	default:
		return false
	}
}

// PreviousScheduleDate returns the calendar day before date (YYYY-MM-DD).
func PreviousScheduleDate(date string) string {
	t, err := time.Parse("2006-01-02", date)
	if err != nil {
		return date
	}
	return t.AddDate(0, 0, -1).Format("2006-01-02")
}

// FetchLiveGamePKs returns game primary keys that are currently Live according
// to the MLB schedule endpoint. Checks today and yesterday's ET slates so
// west-coast games after midnight ET are still discovered.
func FetchLiveGamePKs(ctx context.Context, httpClient *http.Client, date string) ([]int, error) {
	if date == "" {
		date = ScheduleDateET(time.Now())
	}

	dates := []string{PreviousScheduleDate(date), date}
	seen := make(map[int]struct{})
	var pks []int

	for _, scheduleDate := range dates {
		games, err := FetchScheduleGames(ctx, httpClient, scheduleDate)
		if err != nil {
			return nil, err
		}
		for _, game := range games {
			if !IsLiveStatus(game.Status) {
				continue
			}
			if _, ok := seen[game.GamePK]; ok {
				continue
			}
			seen[game.GamePK] = struct{}{}
			pks = append(pks, game.GamePK)
		}
	}

	return pks, nil
}

// DiscoverLiveGamePKs fetches currently live game IDs using this client's HTTP pool.
func (c *Client) DiscoverLiveGamePKs(ctx context.Context, date string) ([]int, error) {
	return FetchLiveGamePKs(ctx, c.httpClient, date)
}

// FetchScheduleGames loads the day's schedule using this client's HTTP pool.
func (c *Client) FetchScheduleGames(ctx context.Context, date string) ([]ScheduleGame, error) {
	return FetchScheduleGames(ctx, c.httpClient, date)
}

File: web/lib/mlb/schedule.ts
import {
  formatPitchHand,
  formatSeasonPitcherLine,
} from "@/lib/mlb/cardPitchers";
import {
  LIVE_GAME_STATUSES,
  TRACKED_GAME_STATUSES,
  type ActiveGame,
  type CardPitcher,
  type MLBScheduleGame,
  type MLBScheduleResponse,
  type SlateGame,
} from "@/types/mlb";

import { cachedScheduleFetch } from "@/lib/mlb/scheduleCache";

const MLB_SCHEDULE_BASE = "https://statsapi.mlb.com/api/v1";
const MLB_TIME_ZONE = "America/New_York";

/** MLB schedule dates use US Eastern (league local) calendar days. */
export function getMLBScheduleDate(date = new Date()): string {
  return getCalendarDateInTimeZone(date, MLB_TIME_ZONE);
}

/** Calendar date (YYYY-MM-DD) in an IANA timezone. */
export function getCalendarDateInTimeZone(date: Date, timeZone: string): string {
  return new Intl.DateTimeFormat("en-CA", { timeZone }).format(date);
}

/** Browser-local calendar date (YYYY-MM-DD) for history browsing defaults. */
export function getLocalCalendarDate(date = new Date()): string {
  return getCalendarDateInTimeZone(date, getBrowserTimeZone());
}

/** Resolved browser IANA timezone (client-only; falls back to Eastern on the server). */
export function getBrowserTimeZone(): string {
  if (typeof Intl === "undefined") return MLB_TIME_ZONE;
  return Intl.DateTimeFormat().resolvedOptions().timeZone || MLB_TIME_ZONE;
}

/** Which local calendar day a scheduled first pitch falls on. */
export function gameLocalCalendarDate(gameDateIso: string, timeZone: string): string {
  return getCalendarDateInTimeZone(new Date(gameDateIso), timeZone);
}

export function addScheduleDays(date: string, days: number): string {
  const [year, month, day] = date.split("-").map(Number);
  const next = new Date(Date.UTC(year, month - 1, day));
  next.setUTCDate(next.getUTCDate() + days);
  return next.toISOString().slice(0, 10);
}

const CARRYOVER_STATUSES = new Set([
  "Live",
  "In Progress",
  "Warmup",
  "Pre-Game",
  "Delayed",
]);

/** Preview games only appear when first pitch is within this window. */
const PREVIEW_WINDOW_MS = 4 * 60 * 60 * 1000;

function toActiveGame(game: MLBScheduleGame): ActiveGame {
  const away = game.teams.away.team.name;
  const home = game.teams.home.team.name;

  return {
    gamePk: game.gamePk,
    awayTeam: away,
    homeTeam: home,
    label: `${away} @ ${home}`,
    status: game.status.abstractGameState,
    gameDate: game.gameDate,
  };
}

function toSlateGame(
  game: MLBScheduleGame,
  pitcherStats: Map<number, PitcherSeasonSnapshot>,
): SlateGame {
  const base = toActiveGame(game);
  const linescore = game.linescore;
  const awayTeam = game.teams.away;
  const homeTeam = game.teams.home;

  return {
    ...base,
    awayAbbrev:
      awayTeam.team.abbreviation ?? awayTeam.team.name.slice(0, 3).toUpperCase(),
    homeAbbrev:
      homeTeam.team.abbreviation ?? homeTeam.team.name.slice(0, 3).toUpperCase(),
    awayScore: awayTeam.score ?? linescore?.teams?.away?.runs ?? null,
    homeScore: homeTeam.score ?? linescore?.teams?.home?.runs ?? null,
    awayHits: linescore?.teams?.away?.hits ?? null,
    homeHits: linescore?.teams?.home?.hits ?? null,
    awayErrors: linescore?.teams?.away?.errors ?? null,
    homeErrors: linescore?.teams?.home?.errors ?? null,
    currentInning: linescore?.currentInning ?? null,
    inningHalf: linescore?.inningHalf ?? linescore?.inningState ?? null,
    inningState: linescore?.inningState ?? null,
    awayPitcher: probableCardPitcher(awayTeam.probablePitcher, pitcherStats),
    homePitcher: probableCardPitcher(homeTeam.probablePitcher, pitcherStats),
  };
}

interface PitcherSeasonSnapshot {
  throwHand: string | null;
  wins: number | null;
  losses: number | null;
  era: string | null;
}

function probableCardPitcher(
  probable: MLBScheduleGame["teams"]["away"]["probablePitcher"],
  stats: Map<number, PitcherSeasonSnapshot>,
): CardPitcher | null {
  const playerId = probable?.id;
  const name = probable?.fullName?.trim();
  if (!playerId || !name) return null;

  const season = stats.get(playerId);
  const throwHand =
    formatPitchHand(probable?.pitchHand?.code) ?? season?.throwHand ?? null;

  return {
    playerId,
    name,
    throwHand,
    line: formatSeasonPitcherLine(
      season?.wins ?? null,
      season?.losses ?? null,
      season?.era ?? null,
    ),
  };
}

async function fetchPitcherSeasonStats(
  personIds: number[],
): Promise<Map<number, PitcherSeasonSnapshot>> {
  const unique = [...new Set(personIds.filter((id) => id > 0))];
  if (unique.length === 0) return new Map();

  const url = new URL(`${MLB_SCHEDULE_BASE}/people`);
  url.searchParams.set("personIds", unique.join(","));
  url.searchParams.set("hydrate", "stats(group=pitching,type=season)");

  const response = await fetch(url.toString(), { cache: "no-store" });
  if (!response.ok) return new Map();

  const data = (await response.json()) as {
    people?: Array<{
      id?: number;
      pitchHand?: { code?: string };
      stats?: Array<{
        splits?: Array<{ stat?: { wins?: number; losses?: number; era?: string } }>;
      }>;
    }>;
  };

  const result = new Map<number, PitcherSeasonSnapshot>();
  for (const person of data.people ?? []) {
    if (!person.id) continue;
    const stat = person.stats?.[0]?.splits?.[0]?.stat;
    result.set(person.id, {
      throwHand: formatPitchHand(person.pitchHand?.code),
      wins: stat?.wins ?? null,
      losses: stat?.losses ?? null,
      era: stat?.era ?? null,
    });
  }

  return result;
}

function statusRank(status: string): number {
  if (LIVE_GAME_STATUSES.has(status)) return 0;
  if (status === "Warmup" || status === "Pre-Game") return 1;
  if (status === "Preview") return 2;
  if (status === "Delayed") return 3;
  return 4;
}

function sortGames(a: ActiveGame, b: ActiveGame): number {
  const rankDiff = statusRank(a.status) - statusRank(b.status);
  if (rankDiff !== 0) return rankDiff;
  return new Date(a.gameDate).getTime() - new Date(b.gameDate).getTime();
}

function isTrackedNow(
  game: MLBScheduleGame,
  options: {
    now?: number;
    localToday?: string;
    timeZone?: string;
    isTodaySlate?: boolean;
  } = {},
): boolean {
  const {
    now = Date.now(),
    localToday,
    timeZone = MLB_TIME_ZONE,
    isTodaySlate = true,
  } = options;
  const status = game.status.abstractGameState;
  if (!TRACKED_GAME_STATUSES.has(status)) return false;

  if (localToday) {
    const gameDay = gameLocalCalendarDate(game.gameDate, timeZone);
    const prevLocalDay = addScheduleDays(localToday, -1);
    const isLocalToday = gameDay === localToday;
    const isPrevLocalCarryover = gameDay === prevLocalDay;

    if (LIVE_GAME_STATUSES.has(status)) {
      return isLocalToday || isPrevLocalCarryover;
    }
    if (status === "Warmup" || status === "Pre-Game" || status === "Delayed") {
      return isLocalToday || isPrevLocalCarryover;
    }
    if (status === "Preview") {
      return isLocalToday;
    }
    return false;
  }

  if (LIVE_GAME_STATUSES.has(status)) return true;
  if (status === "Warmup" || status === "Pre-Game" || status === "Delayed") return true;

  if (status === "Preview") {
    // Legacy ET-slate path: today's full slate is always visible; only apply the
    // preview window to carryover games from yesterday's ET date.
    if (isTodaySlate) return true;
    const startMs = new Date(game.gameDate).getTime();
    return startMs - now <= PREVIEW_WINDOW_MS;
  }

  return false;
}

async function fetchScheduleForDate(date: string): Promise<MLBScheduleGame[]> {
  return cachedScheduleFetch(`schedule:${date}`, async () => {
    const url = new URL(`${MLB_SCHEDULE_BASE}/schedule`);
    url.searchParams.set("sportId", "1");
    url.searchParams.set("date", date);
    url.searchParams.set("gameTypes", "R");
    url.searchParams.set("hydrate", "probablePitcher,linescore,team");

    const response = await fetch(url.toString(), {
      cache: "no-store",
    });

    if (!response.ok) {
      throw new Error(`MLB schedule request failed: ${response.status} ${response.statusText}`);
    }

    const data = (await response.json()) as MLBScheduleResponse;
    return data.dates?.flatMap((d) => d.games ?? []) ?? [];
  });
}

/**
 * Fetches today's regular-season schedule and returns games that are live
 * or about to start. Includes carryover west-coast games from yesterday's ET
 * slate after midnight Eastern, and hides Preview games far in the future.
 */
export async function fetchActiveGames(scheduleDate?: string): Promise<ActiveGame[]> {
  const slate = await fetchSlateGames(scheduleDate);
  return slate;
}

/** Today's slate with linescore summary for game cards. */
export async function fetchSlateGames(
  scheduleDate?: string,
  timeZone?: string,
): Promise<SlateGame[]> {
  const tz = timeZone ?? MLB_TIME_ZONE;
  const localToday = scheduleDate ?? getCalendarDateInTimeZone(new Date(), tz);
  const etToday = getMLBScheduleDate();
  const etDates = [
    addScheduleDays(etToday, -1),
    etToday,
    addScheduleDays(etToday, 1),
  ];

  const byPk = new Map<number, MLBScheduleGame>();
  const slateBatches = await Promise.all(etDates.map((date) => fetchScheduleForDate(date)));
  for (const batch of slateBatches) {
    for (const game of batch) {
      byPk.set(game.gamePk, game);
    }
  }

  const tracked = [...byPk.values()].filter((game) =>
    isTrackedNow(game, { localToday, timeZone: tz }),
  );

  const probableIds = tracked.flatMap((game) => [
    game.teams.away.probablePitcher?.id ?? 0,
    game.teams.home.probablePitcher?.id ?? 0,
  ]);
  const pitcherStats = await fetchPitcherSeasonStats(probableIds);

  return tracked.map((game) => toSlateGame(game, pitcherStats)).sort(sortGames);
}

/** Returns only game PKs with Live / In Progress status (for ingestor parity). */
export async function fetchLiveGamePks(scheduleDate?: string): Promise<number[]> {
  const games = await fetchActiveGames(scheduleDate);
  return games.filter((g) => LIVE_GAME_STATUSES.has(g.status)).map((g) => g.gamePk);
}

/** Previous calendar day for carryover game queries. */
export function previousScheduleDate(date: string): string {
  return addScheduleDays(date, -1);
}

/** Statuses still active on a prior slate that should appear on the next day. */
export const ACTIVE_CARRYOVER_STATUSES = CARRYOVER_STATUSES;

/** Build a Season History URL that preserves browse position. */
export function buildSeasonHistoryHref(options: {
  date?: string;
  view?: "date" | "team";
  teamId?: number | null;
}): string {
  const params = new URLSearchParams();
  if (options.view === "team" && options.teamId) {
    params.set("teamId", String(options.teamId));
    params.set("view", "team");
  } else if (options.date) {
    params.set("date", options.date);
    params.set("view", "date");
  }
  const qs = params.toString();
  return qs ? `/games?${qs}` : "/games";
}

/** Build a game detail URL that carries Season History context for back navigation. */
export function buildGameDetailHref(
  gamePk: number,
  history?: { date?: string; view?: "date" | "team"; teamId?: number | null },
): string {
  const params = new URLSearchParams();
  if (history?.view === "team" && history.teamId) {
    params.set("teamId", String(history.teamId));
    params.set("view", "team");
  } else if (history?.date) {
    params.set("date", history.date);
    params.set("view", "date");
  }
  const qs = params.toString();
  return qs ? `/games/${gamePk}?${qs}` : `/games/${gamePk}`;
}
